// @enderfall/ui/src/index.ts
export type ThemeMode = string;
export { AccessGate } from "./components/AccessGate";
export { Button } from "./components/Button";
export { FormField } from "./components/FormField";
export { Input, Textarea, Select } from "./components/Input";
export { Panel } from "./components/Panel";
export { StackedCard } from "./components/StackedCard";
export { Dropdown } from "./components/Dropdown";
export { MainHeader } from "./components/MainHeader";
export { Modal } from "./components/Modal";
export { PreferencesModal } from "./components/PreferencesModal";
export { Toggle } from "./components/Toggle";
export { Slider } from "./components/Slider";
export { StatDots } from "./components/StatDots";
export { SideMenu, SideMenuSubmenu } from "./components/SideMenu";

type ThemeOptions<T extends ThemeMode> = {
  storageKey: string;
  defaultTheme: T;
  allowed: readonly T[];
  dataAttribute?: string;
  bodyClass?: string | null;
};

export const getStoredTheme = <T extends ThemeMode>(options: ThemeOptions<T>) => {
  const stored = localStorage.getItem(options.storageKey);
  if (stored && options.allowed.includes(stored as T)) {
    return stored as T;
  }
  if (stored === "dark" && options.allowed.includes("galaxy" as T)) {
    return "galaxy" as T;
  }
  return options.defaultTheme;
};

export const applyTheme = <T extends ThemeMode>(theme: T, options: ThemeOptions<T>) => {
  const dataAttribute = options.dataAttribute ?? "data-theme";
  if (options.bodyClass) {
    document.body.classList.toggle(options.bodyClass, theme === (options.defaultTheme as string));
  }
  document.documentElement.setAttribute(dataAttribute, theme);
  localStorage.setItem(options.storageKey, theme);
};
