// @enderfall/ui/src/components/AccessGate.tsx
import { Button } from "./Button";
import { Panel } from "./Panel";

type AccessGateProps = {
  status: "checking" | "locked" | "allowed";
  titleChecking?: string;
  titleLocked?: string;
  messageChecking?: string;
  messageLocked?: string;
  primaryLabel: string;
  secondaryLabel?: string;
  onPrimary: () => void;
  onSecondary?: () => void;
  primaryClassName?: string;
  secondaryClassName?: string;
};

export const AccessGate = ({
  status,
  titleChecking = "Checking access...",
  titleLocked = "App locked",
  messageChecking = "Verifying your access with Enderfall Hub.",
  messageLocked = "Open Enderfall Hub to verify premium or admin access.",
  primaryLabel,
  secondaryLabel,
  onPrimary,
  onSecondary,
  primaryClassName = "primary",
  secondaryClassName = "ghost",
}: AccessGateProps) => {
  if (status === "allowed") return null;
  const isChecking = status === "checking";
  return (
    <div className="ef-access-overlay">
      <Panel variant="card" borderWidth={2} className="ef-access-card">
        <div className="ef-access-title">{isChecking ? titleChecking : titleLocked}</div>
        <p>{isChecking ? messageChecking : messageLocked}</p>
        <div className="ef-access-actions">
          <Button
            type="button"
            className={primaryClassName}
            onClick={onPrimary}
            disabled={isChecking}
          >
            {primaryLabel}
          </Button>
          {secondaryLabel && onSecondary ? (
            <Button
              type="button"
              className={secondaryClassName}
              onClick={onSecondary}
              disabled={isChecking}
            >
              {secondaryLabel}
            </Button>
          ) : null}
        </div>
      </Panel>
    </div>
  );
};
