// @enderfall/ui/src/components/MainHeader.tsx
import { useRef, type ReactNode } from "react";
import type { HeaderMenuItem } from "./Dropdown";
import { Button } from "./Button";
import { Panel } from "./Panel";

type MainHeaderProps = {
  title?: string;
  subtitle?: string;
  logoSrc?: string;
  menus: HeaderMenuItem[];
  menuOpen: string | null;
  onOpenMenu: (id: string) => void;
  onCloseMenu: () => void;
  actions?: ReactNode;
};

export const MainHeader = ({
  title = "Enderfall",
  subtitle = "Galaxy tools for creators",
  logoSrc,
  menus,
  menuOpen,
  onOpenMenu,
  onCloseMenu,
  actions,
}: MainHeaderProps) => {
  const closeTimerRef = useRef<number | null>(null);
  const closeDelayMs = 160;

  const cancelScheduledClose = () => {
    if (closeTimerRef.current !== null) {
      window.clearTimeout(closeTimerRef.current);
      closeTimerRef.current = null;
    }
  };

  const scheduleClose = () => {
    cancelScheduledClose();
    closeTimerRef.current = window.setTimeout(() => {
      onCloseMenu();
    }, closeDelayMs);
  };

  return (
    <Panel variant="header" borderWidth={2} className="ef-main-header">
      <div className="ef-header-left">
        <div className="ef-brand">
          {logoSrc ? (
            <img src={logoSrc} alt={title} className="ef-logo" />
          ) : (
            <div className="ef-logo-fallback">E</div>
          )}
          <div>
            <div className="ef-brand-name">{title}</div>
            {subtitle ? <div className="ef-tagline">{subtitle}</div> : null}
          </div>
        </div>
        <div className="ef-menu-bar">
          {menus.map((menu) => (
            <div
              key={menu.id}
              className="ef-menu-group"
              data-open={menuOpen === menu.id ? "true" : "false"}
              onMouseEnter={() => {
                cancelScheduledClose();
                onOpenMenu(menu.id);
              }}
              onMouseLeave={scheduleClose}
            >
              <Button
                className={["ef-menu-button", "ef-tab", menuOpen === menu.id ? "is-active" : ""]
                  .filter(Boolean)
                  .join(" ")}
                type="button"
                variant="tab"
                subvariant="default"
                data-open={menuOpen === menu.id ? "true" : "false"}
              >
                {menu.label}
              </Button>
              {menuOpen === menu.id ? (
                <div
                  className="ef-menu-popover"
                  data-open="true"
                  onMouseEnter={() => {
                    cancelScheduledClose();
                    onOpenMenu(menu.id);
                  }}
                  onMouseLeave={scheduleClose}
                >
                  {menu.content}
                </div>
              ) : null}
            </div>
          ))}
        </div>
      </div>
      <div className="ef-header-actions">{actions}</div>
    </Panel>
  );
};
