// @enderfall/ui/src/components/MainHeader.tsx
import type { ReactNode } from "react";
import { Dropdown, type HeaderMenuItem } from "./Dropdown";
import { Panel } from "./Panel";

type MainHeaderProps = {
  title?: string;
  subtitle?: string;
  logoSrc?: string;
  menus: HeaderMenuItem[];
  menuOpen: string | null;
  onOpenMenu: (id: string) => void;
  onCloseMenu: () => void;
  actions?: ReactNode;
};

export const MainHeader = ({
  title = "Enderfall",
  subtitle = "Galaxy tools for creators",
  logoSrc,
  menus,
  menuOpen,
  onOpenMenu,
  onCloseMenu,
  actions,
}: MainHeaderProps) => {
  return (
    <Panel variant="header" borderWidth={2} className="ef-main-header">
      <div className="ef-header-left">
        <div className="ef-brand">
          {logoSrc ? (
            <img src={logoSrc} alt={title} className="ef-logo" />
          ) : (
            <div className="ef-logo-fallback">E</div>
          )}
          <div>
            <div className="ef-brand-name">{title}</div>
            {subtitle ? <div className="ef-tagline">{subtitle}</div> : null}
          </div>
        </div>
        <Dropdown
          variant="header"
          menus={menus}
          menuOpen={menuOpen}
          onOpenMenu={onOpenMenu}
          onCloseMenu={onCloseMenu}
        />
      </div>
      <div className="ef-header-actions">{actions}</div>
    </Panel>
  );
};
