// @enderfall/ui/src/components/Modal.tsx
import { useEffect, useRef, useState, type CSSProperties, type ReactNode } from "react";
import { Panel } from "./Panel";

type ModalSize = "default" | "wide";

const MODAL_TRANSITION_MS = 240;

type ModalProps = {
  isOpen: boolean;
  title: string;
  subtitle?: string;
  size?: ModalSize;
  onClose: () => void;
  actions?: ReactNode;
  headerActions?: ReactNode;
  className?: string;
  children?: ReactNode;
};

export const Modal = ({
  isOpen,
  title,
  subtitle,
  size = "default",
  onClose,
  actions,
  headerActions,
  className,
  children,
}: ModalProps) => {
  const reduceMotion =
    typeof document !== "undefined" && document.documentElement.getAttribute("data-reduce-motion") === "true";
  const transitionMs = reduceMotion ? 0 : MODAL_TRANSITION_MS;
  const [isRendered, setIsRendered] = useState(isOpen);
  const [isEntering, setIsEntering] = useState(false);
  const [isExiting, setIsExiting] = useState(false);
  const [enterOffset, setEnterOffset] = useState<{ x: number; y: number }>({ x: 0, y: 0 });
  const exitTimerRef = useRef<number | null>(null);
  const enterFrameRef = useRef<number | null>(null);

  useEffect(() => {
    return () => {
      if (exitTimerRef.current !== null) window.clearTimeout(exitTimerRef.current);
      if (enterFrameRef.current !== null) window.cancelAnimationFrame(enterFrameRef.current);
    };
  }, []);

  useEffect(() => {
    if (isOpen) {
      if (exitTimerRef.current !== null) {
        window.clearTimeout(exitTimerRef.current);
        exitTimerRef.current = null;
      }

      const activeElement = document.activeElement as HTMLElement | null;
      const activeRect = activeElement?.getBoundingClientRect();
      const sourceX = activeRect ? activeRect.left + activeRect.width / 2 : window.innerWidth / 2;
      const sourceY = activeRect ? activeRect.top + activeRect.height / 2 : window.innerHeight - 56;
      const centerX = window.innerWidth / 2;
      const centerY = window.innerHeight / 2;

      setEnterOffset({
        x: (sourceX - centerX) * 0.18,
        y: (sourceY - centerY) * 0.22,
      });

      setIsRendered(true);
      setIsExiting(false);
      setIsEntering(!reduceMotion);

      if (reduceMotion) return;

      if (enterFrameRef.current !== null) window.cancelAnimationFrame(enterFrameRef.current);
      enterFrameRef.current = window.requestAnimationFrame(() => {
        enterFrameRef.current = window.requestAnimationFrame(() => {
          setIsEntering(false);
          enterFrameRef.current = null;
        });
      });
      return;
    }

    if (!isRendered) return;
    if (reduceMotion) {
      setIsEntering(false);
      setIsExiting(false);
      setIsRendered(false);
      return;
    }

    setIsEntering(false);
    setIsExiting(true);

    if (exitTimerRef.current !== null) window.clearTimeout(exitTimerRef.current);
    exitTimerRef.current = window.setTimeout(() => {
      setIsRendered(false);
      setIsExiting(false);
      exitTimerRef.current = null;
    }, transitionMs);
  }, [isOpen, isRendered, reduceMotion, transitionMs]);

  if (!isRendered) return null;

  const sizeClass = size === "wide" ? "ef-modal--wide" : "";
  const modalStateClass = isExiting ? "is-exiting" : isEntering ? "is-entering" : "is-open";
  const backdropClasses = ["ef-modal-backdrop", modalStateClass].filter(Boolean).join(" ");
  const classes = ["ef-modal", sizeClass, modalStateClass, className].filter(Boolean).join(" ");
  const modalStyle = {
    "--ef-modal-enter-x": `${enterOffset.x}px`,
    "--ef-modal-enter-y": `${enterOffset.y}px`,
  } as CSSProperties;

  return (
    <div className={backdropClasses} onClick={onClose}>
      <Panel
        variant="card"
        borderWidth={2}
        className={classes}
        style={modalStyle}
        onClick={(event) => event.stopPropagation()}
      >
        <div className="ef-modal-header">
          <div className="ef-modal-title">{title}</div>
          <div className="ef-modal-header-actions">
            {headerActions}
            <button
              className="icon-action small ef-modal-close"
              onClick={onClose}
              type="button"
              aria-label="Close"
            >
              {"\u00D7"}
            </button>
          </div>
        </div>
        {subtitle ? <div className="ef-modal-subtitle">{subtitle}</div> : null}
        {children ? <div className="ef-modal-body">{children}</div> : null}
        {actions ? <div className="ef-modal-actions">{actions}</div> : null}
      </Panel>
    </div>
  );
};
