// @enderfall/ui/src/components/Modal.tsx
import type { ReactNode } from "react";
import { Panel } from "./Panel";

type ModalSize = "default" | "wide";

type ModalProps = {
  isOpen: boolean;
  title: string;
  subtitle?: string;
  size?: ModalSize;
  onClose: () => void;
  actions?: ReactNode;
  className?: string;
  children?: ReactNode;
};

export const Modal = ({
  isOpen,
  title,
  subtitle,
  size = "default",
  onClose,
  actions,
  className,
  children,
}: ModalProps) => {
  if (!isOpen) return null;
  const sizeClass = size === "wide" ? "ef-modal--wide" : "";
  const classes = ["ef-modal", sizeClass, className].filter(Boolean).join(" ");
  return (
    <div className="ef-modal-backdrop" onClick={onClose}>
      <Panel
        variant="card"
        borderWidth={2}
        className={classes}
        onClick={(event) => event.stopPropagation()}
      >
        <div className="ef-modal-header">
          <div className="ef-modal-title">{title}</div>
          <button
            className="icon-action small ef-modal-close"
            onClick={onClose}
            type="button"
            aria-label="Close"
          >
            ×
          </button>
        </div>
        {subtitle ? <div className="ef-modal-subtitle">{subtitle}</div> : null}
        {children ? <div className="ef-modal-body">{children}</div> : null}
        {actions ? <div className="ef-modal-actions">{actions}</div> : null}
      </Panel>
    </div>
  );
};
