// @enderfall/ui/src/components/PreferencesModal.tsx
import type { ReactNode } from "react";
import { Button } from "./Button";
import { Dropdown } from "./Dropdown";
import { Toggle } from "./Toggle";
import { Modal } from "./Modal";

type ThemeOption = { value: string; label: string };

type PreferencesModalProps = {
  isOpen: boolean;
  onClose: () => void;
  themeMode: string;
  onThemeChange: (value: string) => void;
  themeOptions: ThemeOption[];
  animationsEnabled: boolean;
  onAnimationsChange: (enabled: boolean) => void;
  note?: string;
  children?: ReactNode;
};

export const PreferencesModal = ({
  isOpen,
  onClose,
  themeMode,
  onThemeChange,
  themeOptions,
  animationsEnabled,
  onAnimationsChange,
  note = "Syncs across all Enderfall apps.",
  children,
}: PreferencesModalProps) => {
  const selectedTheme =
    themeOptions.find((option) => option.value === themeMode) ?? themeOptions[0];
  const dropdownLabel = selectedTheme?.label ?? "Select theme";
  const dropdownItems = themeOptions.map((option) => ({
    id: option.value,
    label: option.label,
    onClick: () => onThemeChange(option.value),
    className: `theme-preview theme-preview--${option.value}`,
    variant: "theme-preview" as const,
  }));

  return (
    <Modal isOpen={isOpen} onClose={onClose} title="Preferences" subtitle="Applies across Enderfall apps.">
      <div className="ef-modal-form">
        <label>
          Theme
          <div className="ef-modal-user-dropdown">
            <Dropdown variant="user" name={dropdownLabel} items={dropdownItems} />
          </div>
        </label>
        <Toggle
          checked={animationsEnabled}
          onChange={(event) => onAnimationsChange(event.target.checked)}
          label="Enable animations"
        />
        {children}
        {note ? <div className="ef-modal-note">{note}</div> : null}
      </div>
      <div className="ef-modal-actions">
        <Button variant="primary" type="button" onClick={onClose}>
          Close
        </Button>
      </div>
    </Modal>
  );
};
