// @enderfall/ui/src/components/SideMenu.tsx
import {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
  type CSSProperties,
  type ReactNode,
  type MouseEvent,
} from "react";

type SideMenuContextValue = {
  openId: string | null;
  setOpenId: (id: string | null) => void;
  scheduleOpen: (id: string) => void;
  clearHover: () => void;
  toggleOpen: (id: string) => void;
  openedAt: number;
  initialDelayMs: number;
};

const SideMenuContext = createContext<SideMenuContextValue | null>(null);

type SideMenuProps = {
  children: ReactNode;
  resetKey?: string | number | boolean | null;
  hoverDelayMs?: number;
  initialDelayMs?: number;
  onOpenChange?: (openId: string | null) => void;
};

export const SideMenu = ({
  children,
  resetKey,
  hoverDelayMs = 220,
  initialDelayMs = 180,
  onOpenChange,
}: SideMenuProps) => {
  const [openId, setOpenId] = useState<string | null>(null);
  const hoverTimerRef = useRef<number | null>(null);
  const openedAtRef = useRef<number>(Date.now());

  useEffect(() => {
    if (hoverTimerRef.current !== null) {
      window.clearTimeout(hoverTimerRef.current);
      hoverTimerRef.current = null;
    }
    setOpenId(null);
    openedAtRef.current = Date.now();
  }, [resetKey]);

  useEffect(() => {
    onOpenChange?.(openId);
  }, [openId, onOpenChange]);

  useEffect(() => {
    return () => {
      if (hoverTimerRef.current !== null) {
        window.clearTimeout(hoverTimerRef.current);
      }
    };
  }, []);

  const scheduleOpen = (id: string) => {
    if (hoverTimerRef.current !== null) {
      window.clearTimeout(hoverTimerRef.current);
    }
    const elapsed = Date.now() - openedAtRef.current;
    const settleDelay = elapsed < initialDelayMs ? initialDelayMs - elapsed : 0;
    hoverTimerRef.current = window.setTimeout(() => {
      setOpenId(id);
    }, hoverDelayMs + settleDelay);
  };

  const clearHover = () => {
    if (hoverTimerRef.current !== null) {
      window.clearTimeout(hoverTimerRef.current);
      hoverTimerRef.current = null;
    }
  };

  const toggleOpen = (id: string) => {
    setOpenId((prev) => (prev === id ? null : id));
  };

  const value = useMemo(
    () => ({
      openId,
      setOpenId,
      scheduleOpen,
      clearHover,
      toggleOpen,
      openedAt: openedAtRef.current,
      initialDelayMs,
    }),
    [openId, initialDelayMs]
  );

  return <SideMenuContext.Provider value={value}>{children}</SideMenuContext.Provider>;
};

type SideMenuTriggerProps = {
  onClick: (event: MouseEvent) => void;
  disabled?: boolean;
  "aria-expanded": boolean;
};

type SideMenuSubmenuProps = {
  id: string;
  trigger: (props: SideMenuTriggerProps) => ReactNode;
  children: ReactNode;
  className?: string;
  panelClassName?: string;
  disabled?: boolean;
  enableViewportFlip?: boolean;
  onOpenChange?: (open: boolean) => void;
  variant?: "default" | "header";
};

export const SideMenuSubmenu = ({
  id,
  trigger,
  children,
  className,
  panelClassName,
  disabled,
  enableViewportFlip = false,
  onOpenChange,
  variant = "default",
}: SideMenuSubmenuProps) => {
  const context = useContext(SideMenuContext);
  if (!context) {
    throw new Error("SideMenuSubmenu must be used within a SideMenu provider.");
  }

  const { openId, setOpenId, scheduleOpen, clearHover, toggleOpen, openedAt, initialDelayMs } =
    context;
  const isOpen = openId === id;
  const panelRef = useRef<HTMLDivElement | null>(null);
  const containerRef = useRef<HTMLDivElement | null>(null);
  const [panelStyle, setPanelStyle] = useState<CSSProperties>({});

  useEffect(() => {
    onOpenChange?.(isOpen);
  }, [isOpen, onOpenChange]);

  useEffect(() => {
    if (!isOpen || !enableViewportFlip) {
      setPanelStyle({});
      return;
    }
    const updatePosition = () => {
      const panel = panelRef.current;
      const container = containerRef.current;
      if (!panel || !container) return;
      const panelRect = panel.getBoundingClientRect();
      const containerRect = container.getBoundingClientRect();
      const margin = 12;
      const defaultLeft = containerRect.right + 6;
      const flippedLeft = containerRect.left - panelRect.width - 6;
      let left = defaultLeft;
      if (defaultLeft + panelRect.width > window.innerWidth - margin && flippedLeft >= margin) {
        left = flippedLeft;
      }
      left = Math.min(Math.max(left, margin), window.innerWidth - margin - panelRect.width);
      let top = containerRect.top;
      top = Math.min(Math.max(top, margin), window.innerHeight - margin - panelRect.height);
      setPanelStyle({
        left: left - containerRect.left,
        top: top - containerRect.top,
        right: "auto",
      });
    };
    const raf = requestAnimationFrame(updatePosition);
    window.addEventListener("resize", updatePosition);
    window.addEventListener("scroll", updatePosition, true);
    return () => {
      cancelAnimationFrame(raf);
      window.removeEventListener("resize", updatePosition);
      window.removeEventListener("scroll", updatePosition, true);
    };
  }, [isOpen, enableViewportFlip]);

  const handlePointerEnter = () => {
    if (disabled || isOpen) return;
    if (Date.now() - openedAt < initialDelayMs) return;
    setOpenId(null);
    scheduleOpen(id);
  };

  const handlePointerLeave = () => {
    clearHover();
  };

  const handlePointerMove = () => {
    if (disabled || isOpen) return;
    if (Date.now() - openedAt < initialDelayMs) return;
    scheduleOpen(id);
  };

  const handleClick = (event: MouseEvent) => {
    if (disabled) return;
    event.preventDefault();
    if (isOpen) return;
    setOpenId(id);
  };

  return (
    <div
      className={[
        className,
        isOpen ? "is-open" : "",
        variant === "header" ? "side-menu--header" : "",
      ]
        .filter(Boolean)
        .join(" ")}
      data-submenu-id={id}
      ref={containerRef}
      onPointerEnter={handlePointerEnter}
      onPointerLeave={handlePointerLeave}
      onPointerMove={handlePointerMove}
    >
      {trigger({ onClick: handleClick, disabled, "aria-expanded": isOpen })}
      <div
        ref={panelRef}
        className={[
          panelClassName,
          isOpen ? "is-open" : "",
        ]
          .filter(Boolean)
          .join(" ")}
        style={panelStyle}
      >
        {children}
      </div>
    </div>
  );
};
