// @enderfall/ui/src/components/StackedCard.tsx
import type { CSSProperties, ReactNode } from "react";

type Variant = "games" | "mods" | "servers" | "apps";
type Align = "center" | "left";
type Tone = "default" | "plain";

type Action = {
  label: string;
  href?: string;
  onClick?: () => void;
};

type StackedCardProps = {
  title?: string;
  description?: string;
  tags?: string[];
  imageUrl?: string | null;
  variant?: Variant;
  action?: Action;
  align?: Align;
  tone?: Tone;
  showImage?: boolean;
  children?: ReactNode;
  frameClassName?: string;
  cardClassName?: string;
  bodyClassName?: string;
  id?: string;
  style?: CSSProperties;
};

const variantClass = (variant?: Variant) => {
  switch (variant) {
    case "games":
      return "ef-stacked-variant-games";
    case "mods":
      return "ef-stacked-variant-mods";
    case "servers":
      return "ef-stacked-variant-servers";
    case "apps":
      return "ef-stacked-variant-apps";
    default:
      return "";
  }
};

const alignClass = (align?: Align) => (align === "left" ? "ef-stacked-body-left" : "");
const toneClass = (tone?: Tone) => (tone === "plain" ? "ef-stacked-body-plain" : "ef-stacked-body");

export const StackedCard = ({
  title = "",
  description = "",
  tags,
  imageUrl,
  variant,
  action,
  align = "center",
  tone = "default",
  showImage = true,
  children,
  frameClassName,
  cardClassName,
  bodyClassName,
  id,
  style,
}: StackedCardProps) => {
  const bodyClasses = [toneClass(tone), alignClass(align), bodyClassName]
    .filter(Boolean)
    .join(" ");
  return (
    <article
      className={["ef-stacked-frame", variantClass(variant), frameClassName].filter(Boolean).join(" ")}
      id={id}
      style={style}
    >
      <div className={["ef-stacked-card", cardClassName].filter(Boolean).join(" ")}>
        {showImage ? (
          <div
            className={["ef-stacked-image", imageUrl ? "ef-stacked-image-photo" : ""]
              .filter(Boolean)
              .join(" ")}
            style={
              imageUrl
                ? {
                    backgroundImage: `linear-gradient(180deg, rgba(10, 12, 22, 0.1), rgba(10, 12, 22, 0.85)), url(${imageUrl})`,
                    backgroundSize: "cover",
                    backgroundPosition: "center",
                  }
                : undefined
            }
          />
        ) : null}
        <div className={bodyClasses}>
          {children ? (
            children
          ) : (
            <>
              <h3>{title}</h3>
              <p>{description}</p>
              {tags?.length ? (
                <div className={`ef-stacked-tags ${align === "left" ? "ef-stacked-tags-left" : ""}`}>
                  {tags.map((tag) => (
                    <span key={tag} className="ef-stacked-tag">
                      {tag}
                    </span>
                  ))}
                </div>
              ) : null}
            </>
          )}
          {action ? (
            action.href ? (
              <a className="ef-stacked-action" href={action.href}>
                {action.label}
              </a>
            ) : (
              <button className="ef-stacked-action" type="button" onClick={action.onClick}>
                {action.label}
              </button>
            )
          ) : null}
        </div>
      </div>
    </article>
  );
};
